### PR TITLE
cmd: fix `files ls` to report hash and size for files

### PR DESCRIPTION
**Pull request by schomatis**
Fixes https://github.com/ipfs/go-ipfs/issues/5044.

### CODE REVIEW

**Comment by schomatis** [comment]
I'll add some tests (or probably some already existing tests should be fixed).

**Comment by schomatis** [comment]
The test is failing because the reported size varies for reasons I don't quite understand, I'll raise a different issue for that and leave this as blocked in the meanwhile.

**Comment by Stebalien** [comment]
I wonder if we need fo flush first or something...

**Comment by schomatis** [comment]
Now that you mention it, maybe we need to lock the directory, that's what `ForEachEntry` does when it gets the node (that's where I copied the patch from),

https://github.com/ipfs/go-ipfs/blob/1e9e2f453c435272433a96ebfba5c4319cc08534/mfs/dir.go#L262-L274

(but that may just be due to the `childUnsync` call, I can't tell).

**Comment by Stebalien** [comment]
Sorry, I missed that commit before my comment. It looks like you fixed the size issue, right? The code looks corect, at least.

File: core/commands/files.go [hunk 1]
```diff
@@ -462,7 +462,24 @@ Examples:
...
+
+			nd, err := fsn.GetNode()
+			if err != nil {
+				res.SetError(err, cmdkit.ErrNormal)
+				return
+			}
+
+			out := &filesLsOutput{[]mfs.NodeListing{mfs.NodeListing{
+				Name: name,
+				Type: 1,
```
Comment (Stebalien) [inline core/commands/files.go]: Should probably use `unixfs.TFile` to avoid magic numbers.

Comment (schomatis) [reply core/commands/files.go]: Yes, I should directly call the `Type()` function, `1` isn't even the actual value of `TFile`.., thanks!

File: test/sharness/t0250-files-api.sh [hunk 2]
```diff
@@ -186,6 +186,12 @@ test_files_api() {
     test_cmp ls_l_expected ls_l_actual
   '
 
+  test_expect_success "file has correct hash and size listed with -l" '
+    echo "file1	$FILE1	4" > ls_l_expected &&
+    ipfs files ls -l /cats/file1 > ls_l_actual &&
+    test_cmp ls_l_expected ls_l_actual
+  '
+
```
Comment (kevina) [inline test/sharness/t0250-files-api.sh]: Please add a test for the ls output without `-l` is unchanged when used on a file.

**Comment by kevina** [comment]
> @kevina The field of the NodeListing structure displayed when -l isn't used is just Name,

Okay the reason I bring it up is that 'ls /afile' should not fail if the hash for /afile is not available.  But a quick test on the exiting version of IPFS verified this is not the case, so this is not a regression but something that could be cleaned up later.

so LGTM.

**Comment by schomatis** [comment]
> 'ls /afile' should not fail if the hash for /afile is not available.

Good point, I should be replicating the logic of the `*mfs.Directory` case were that information is loaded only if `long` is true (`-l` option), let me fix that and get back to you.

**Comment by schomatis** [comment]
Fixed, @kevina could you take another look please?

Just to clarify, in the file case, the node has already been fetched through MFS, so getting its hash (or any other information) won't fail, in contrast, in the directory case, after the directory node is fetched other nodes need to be looked in the DAG (files inside the directory) for the command to build its output, leaving open the door for an potential error.

However, even if this case can't fail, it should do the amount of work it needs to do and not more, so it needs to also (as the directory case) discriminate the `-l` true/false cases.

**Comment by schomatis** [comment]
Oh, fun fact, I've just realized that the `Type` field that was hard-coded isn't even needed, I'm still leaving it in the PR just in case future version do show it (it's cheap to compute).

**Pull request merged**